### ISSUE TITLE
Maximally interoperable GeoJSON

### ISSUE THREAD

**Issue by sgillies** [geojson/draft-geojson#74]
We've agreed to **not recommend** use of coordinate systems other than CRS84, which is a big step toward interoperability. It's not unlike the UTF-8 encoding constraint in [I-JSON](http://tools.ietf.org/html/rfc7493#section-2.1): RFC 7159 JSON allows UTF-16 and UTF-32, but I-JSON says "use only UTF-8 for maximum interop."

## I-JSON constraints

I-JSON also adds constraints to JSON numbers, saying:

> I-JSON messages SHOULD NOT include numbers that express greater magnitude or
precision than an IEEE 754 double precision number provides, for
example, 1E400 or 3.141592653589793238462643383279.

and adds constraints to object items:

> Objects in I-JSON messages MUST NOT have members with duplicate names.

## Constraining GeoJSON geometries

I propose that GeoJSON interop could be helped by adding a geometry constraint. Consider the  geometry below:

```json
{ "type": "GeometryCollection", 
  "geometries": [
     { "type": "GeometryCollection", 
       "geometries": [ 
         { "type": "Point", "coordinates": [0, 0] } 
       ] } ] }
```

It's functionally equivalent to `{ "type": "Point", "coordinates": [0, 0] }`: if you read the geoms into a simple features implementation (JTS, GEOS, &c), it should say they are equivalent. But it's clear that arbitrarily nested collections are harmful to interoperability.

About geometry collections we could say

1. maximally interoperable geometry collections MUST NOT contain other geometry collections
2. and/or, maximally interoperable GeoJSON SHOULD NOT contain geometry collections

Questions for you all:

1. Do you support or oppose such constraints?
2. Do they belong in the GeoJSON I-D or a subsequent draft (like I-JSON for JSON)?

**Comment by dret** [comment]
i may be wrong here but i thought that one goal of the I-D was to not make existing GeoJSON invalid (unless it is relatively clearly "wrong", whatever exactly that means). if that's the case, then i think there should be a clean separation between the format itself, and I-GeoJSON.
afaict, I-JSON mostly exists because people wanted to document what was creating interop problems in the existing ecosystem. it's not really that somebody thought of a "simpler JSON", but more that people documented things that are known to potentially cause trouble.
i would suggest to maybe follow a similar route with GeoJSON: focus the I-D on cleaning up the language as suggested in a variety of issues. and then maybe in parallel start collecting evidence for GeoJSON that's valid, but still known to cause interop problems. if patterns emerge (i.e., usages that seem to cause problems in a variety of cases), then it certainly would be worthwhile to document them in an I-GeoJSON I-D.

**Comment by tschaub** [comment]
I like the idea of strongly recommending against arbitrarily nested GeometryCollection.  I think this was an oversight in the creating of the spec (at least I don't remember discussion).  This has been a pain point in working with KML.

**Comment by adoyle** [comment]
Would it complicate matters to beef up section 6 "Interoperability Considerations" with a bunch of SHOULD's? And possibly collecting them into a subsection titled I-GeoJSON?

**Comment by jyutzler** [comment]
I'm not entirely sure what the problem is here. If an implementation can handle a GeometryCollection (as it should), then why would there be a problem with a nested GeometryCollection? Is this problem occurring in the wild? I know it isn't the best idea in the world to nest geometry collections (and I'm not sure why someone would do it other than to be a pain) but I feel that this is a solution in search of a problem.

Something that I feel would aid clarity/interoperability of Geometry Collections is to address the last paragraph of https://github.com/geojson/geojson-ld/issues/9#issuecomment-40501315. I have no issue with what @sgillies says here but I think it would be nice to have a clause in the interoperability section saying something like:

> Maximally interoperable GeoJSON clients will treat all elements of a GeometryCollection as part of a whole and will not apply any specialized interpretation or semantic meaning to them.

**Comment by sgillies** [comment]
Great comments, everybody. @jyutzler Yes, it's a good idea to cover this in the interop section, and then like @dret says, save the strong spec language for another doc.

Actions:

- [ ] Add clause as suggested
- [ ] Draft a GeoJSON doc with some deeply nested collections and try it out on various software

**Comment by dret** [comment]
for the I-GeoJSON doc, some aspects of it maybe can be directly taken from I-JSON. maybe even *all* of I-JSON, saying that anything that's I-GeoJSON MUST be I-JSON. and then there may be some additional constraints, such as nested collections. i guess what goes into this list of GeoJSON specific constraints really should be driven by interop problem reports from the field.

**Comment by dret** [comment]
@sgillies just created some nice test cases over at https://github.com/sgillies/geometry-collections

**Comment by sgillies** [comment]
Raised on the list: https://mailarchive.ietf.org/arch/msg/geojson/P1caE1VmjZllYUhhj_fryOfuIIk.

**Comment by dret** [comment]
do the discussions here (about nested geometry collections being legal but not a great thing) also apply to feature collections? it seems to me that for both collection types, it would be useful to say in the definition of these object types that nesting is allowed. (see #147)

**Comment by adoyle** [comment]
Based on the definition of a GeoJSON object in section 2 [1]  and based on the fact that section 2.3 [2] says that a Feature Collection must have a member “features” but is silent on other members, then other members of a Feature Collection are not GeoJSON objects and thus there is no nesting.

The discussion in #147 laments the fact that GeometryCollection doesn’t actually prevent nested geometries, but consensus seems to be that it should not be disallowed due to backwards compatibility but should be discouraged.

[1] https://tools.ietf.org/html/draft-ietf-geojson-00#section-2 <https://tools.ietf.org/html/draft-ietf-geojson-00#section-2>
[2] https://tools.ietf.org/html/draft-ietf-geojson-00#section-2.3 <https://tools.ietf.org/html/draft-ietf-geojson-00#section-2.3>

**Comment by dret** [comment]
On 2016-02-16 16:02, Allan Doyle wrote:
> Based on the definition of a GeoJSON object in section 2 [1] and based
> on the fact that section 2.3 [2] says that a Feature Collection must
> have a member “features” but is silent on other members, then other
> members of a Feature Collection are not GeoJSON objects and thus there
> is no nesting.

then let's make this explicit. too much required exegesis is the mother 
of non-interop.

> The discussion in #147 laments the fact that GeometryCollection doesn’t
> actually prevent nested geometries, but consensus seems to be that it
> should not be disallowed due to backwards compatibility but should be
> discouraged.

then let's make that explicit as well. it's not that hard:

* have a SHOULD NOT making clear that consumers should expect that, but 
producers should not produce it.

* point to the interop section that contains more details and context.

* maybe also add it to #125 as something that was less clear in the 
original spec.

**Comment by adoyle** [comment]
> On Feb 16, 2016, at 11:43 AM, Erik Wilde <notifications@github.com> wrote:
> 
> On 2016-02-16 16:02, Allan Doyle wrote:
> > Based on the definition of a GeoJSON object in section 2 [1] and based
> > on the fact that section 2.3 [2] says that a Feature Collection must
> > have a member “features” but is silent on other members, then other
> > members of a Feature Collection are not GeoJSON objects and thus there
> > is no nesting.
> 
> then let's make this explicit. too much required exegesis is the mother 
> of non-interop.

This is what @sgillies was working on with #142
https://github.com/geojson/draft-geojson/pull/142#issuecomment-183588647 <https://github.com/geojson/draft-geojson/pull/142#issuecomment-183588647>

**Comment by dret** [comment]
On 2016-02-16 17:51, Allan Doyle wrote:
> This is what @sgillies was working on with #142
> https://github.com/geojson/draft-geojson/pull/142#issuecomment-183588647
> <https://github.com/geojson/draft-geojson/pull/142#issuecomment-183588647>

i forgot looking at that. thanks for the pointer. now i am a bit more 
confused, because there's text talking about "Geometry and 
GeometryCollection objects", which seems to imply that these are 
different things. #148 suggests that treating them like that might 
actually make the GeoJSON model more logical, but the intro currently 
defines that a GeometryCollection is a type of Geometry.

**Comment by sgillies** [comment]
@dret I'll try to clear up the confusion. "Geometry" is an abstract type. "GeometryCollection" is one of its concrete sub-types, along with "Point", "MultiPoint", &c. This hierarchy comes from the OGC simple features spec (see https://tools.ietf.org/html/draft-ietf-geojson-00#section-8.2) and is probably worth spelling out in our draft.

**Comment by dret** [comment]
On 2016-02-17 22:58, Sean Gillies wrote:
> @dret <https://github.com/dret> I'll try to clear up the confusion.
> "Geometry" is an abstract type. "GeometryCollection" is one of its
> concrete sub-types, along with "Point", "MultiPoint", &c. This hierarchy
> comes from the OGC simple features spec (see
> https://tools.ietf.org/html/draft-ietf-geojson-00#section-8.2) and is
> probably worth spelling out in our draft.

given the current state, "GeometryCollection" is one concrete type of 
the "Geometry" abstract type, right? i don't think that's great (#148), 
but that's how it is. talking about "Geometry and GeometryCollections 
objects" (#142) thus is a bit redundant and potentially confusing. if 
#148 is not getting adopted, then the language in #142 (and potentially 
in other places as well) should probably be adjusted to just talk about 
"Geometry objects".

**Comment by sgillies** [comment]
@dret I'll stick to "Geometry objects". Definitely the best way to go.

**Comment by dret** [comment]
sounds reasonable. does https://github.com/geojson/draft-geojson/commit/2602a469656ac93a925dcf04ebf567f2d6db5676 look good?
